src/compiler/checker.ts

```diff
@@ -23692,9 +23828,11 @@ namespace ts {
         }
 
         function isExhaustiveSwitchStatement(node: SwitchStatement): boolean {
-            if (!node.possiblyExhaustive) {
-                return false;
-            }
+            const links = getNodeLinks(node);
+            return links.isExhaustive !== undefined ? links.isExhaustive : (links.isExhaustive = computeExhaustiveSwitchStatement(node));
+        }
+
+        function computeExhaustiveSwitchStatement(node: SwitchStatement): boolean {
             if (node.expression.kind === SyntaxKind.TypeOfExpression) {
                 const operandType = getTypeOfExpression((node.expression as TypeOfExpression).expression);
                 // This cast is safe because the switch is possibly exhaustive and does not contain a default case, so there can be no undefined.

```

Comment:
I wonder whether this needs addressing in any way. Previously `node.possiblyExhaustive` was only true when there was no default clause (plus other conditions). Now, we assume that every call to `isExhaustiveSwitchStatement` is guarded by `flow.clauseStart === flow.clauseEnd`, but if someone were to forget that the cast would no longer be safe. Maybe as a change?
```ts
function getSwitchClauseTypeOfWitnesses(switchStatement: SwitchStatement): (string | undefined)[];
function getSwitchClauseTypeOfWitnesses(switchStatement: SwitchStatement, retainDefault: false): string[];
function getSwitchClauseTypeOfWitnesses(switchStatement: SwitchStatement, retainDefault: boolean): (string | undefined)[];
function getSwitchClauseTypeOfWitnesses(switchStatement: SwitchStatement, retainDefault: boolean = true): (string | undefined)[] {
    const witnesses: (string | undefined)[] = [];
    for (const clause of switchStatement.caseBlock.clauses) {
        if (clause.kind === SyntaxKind.CaseClause) {
            if (clause.expression.kind === SyntaxKind.StringLiteral) {
                witnesses.push((clause.expression as StringLiteral).text);
                continue;
            }
            return emptyArray;
        }
        if (retainDefault) witnesses.push(/*explicitDefaultStatement*/ undefined);
    }
    return witnesses;
}
```